binders/kafka-binder/spring-cloud-stream-binder-kafka-reactive/src/main/java/org/springframework/cloud/stream/binder/reactorkafka/ReactorKafkaBinder.java

```diff
@@ -136,35 +138,50 @@ protected MessageProducer createConsumerEndpoint(ConsumerDestination destination
 				this.configurationProperties);
 		Assert.isInstanceOf(RecordMessageConverter.class, converter);
 		ReceiverOptions<Object, Object> opts = ReceiverOptions.create(configs)
-			.addAssignListener(parts -> System.out.println("Assigned: " + parts))
+			.addAssignListener(parts -> logger.info("Assigned: " + parts))
 			.subscription(Collections.singletonList(destination.getName()));
 
-		return new MessageProducerSupport() {
+		class ReactorMessageProducer extends MessageProducerSupport {
 
-			private final KafkaReceiver<Object, Object> receiver = KafkaReceiver.create(opts);
+			private final List<KafkaReceiver<Object, Object>> receivers = new ArrayList<>();
 
-			private volatile Subscription subscription;
+			private final List<Subscription> subscriptions = new ArrayList<>();
+
+			ReactorMessageProducer() {
+				for (int i = 0; i < properties.getConcurrency(); i++) {
+					this.receivers.add(KafkaReceiver.create(opts));
+				}
+			}
 
 			@SuppressWarnings("unchecked")
 			@Override
 			protected void doStart() {
-				Flux<Message<Object>> flux = receiver
-						.receive()
-						.doOnSubscribe(subs -> this.subscription = subs)
-						.map(record -> (Message<Object>) ((RecordMessageConverter) converter)
-								.toMessage(record, null, null, null));
-				subscribeToPublisher(flux);
+				List<Flux<Message<Object>>> fluxes = new ArrayList<>();
+				int concurrency = properties.getConcurrency();
+				for (int i = 0; i < concurrency; i++) {
+					fluxes.add(this.receivers.get(i)
+							.receive()
+							.doOnSubscribe(subs -> this.subscriptions.add(subs))
+							.map(record -> (Message<Object>) ((RecordMessageConverter) converter)
+									.toMessage(record, null, null, null)));
+				}
+				if (concurrency == 1) {
+					subscribeToPublisher(fluxes.get(0));
+				}
+				else {
+					subscribeToPublisher(Flux.merge(fluxes));
```

Comment:
I don't think it's important to retain order since each consumer is consuming from a different partition; they can always group them back together with a `groupBy()`. This way we get max parallelism (if they have a `.parallel()`) or similar.